--- a/mt7615/main.c
+++ b/mt7615/main.c
@@ -851,8 +851,8 @@ mt7615_ampdu_action(struct ieee80211_hw
 		mtxq->send_bar = false;
 		ret = mt7615_mcu_add_tx_ba(dev, params, true);
 		ssn = mt7615_mac_get_sta_tid_sn(dev, msta->wcid.idx, tid);
-		ieee80211_send_bar(vif, sta->addr, tid,
-				   IEEE80211_SN_TO_SEQ(ssn));
+		if (vif->type != NL80211_IFTYPE_STATION)
+			ieee80211_send_bar(vif, sta->addr, tid, IEEE80211_SN_TO_SEQ(ssn));
 		break;
 	case IEEE80211_AMPDU_TX_STOP_FLUSH:
 	case IEEE80211_AMPDU_TX_STOP_FLUSH_CONT:
