Index: mt76-2022-12-22-5b509e80/util.h
===================================================================
--- mt76-2022-12-22-5b509e80.orig/util.h	2023-11-08 08:33:15.404747018 +0200
+++ mt76-2022-12-22-5b509e80/util.h	2023-11-08 08:33:47.000548625 +0200
@@ -65,7 +65,7 @@
 	if (fn)
 		w->fn = fn;
 	w->task = kthread_run(__mt76_worker_fn, w,
-			      "mt76-%s %s", name, dev_name);
+			      "mt76-%s:%s", name, dev_name);
 
 	if (IS_ERR(w->task)) {
 		ret = PTR_ERR(w->task);
